chainer/dataset/convert.py

```diff
@@ -8,6 +9,81 @@
 from chainer.backends import cuda
 
 
+def converter():
+    """Decorator to make a converter function.
+
+    The target converter must be a callable that accepts two positional
+    arguments: a batch and a device, and returns a converted batch.
+
+    The type of the device argument is :class:`chainer.backend.Device`.
+
+    The types and values of the batches (the first argument and the return
+    value) are not specified: they depend on how the converter is used (e.g.
+    by updaters).
+
+    .. admonition:: Example
+
+        >>> @chainer.dataset.converter()
+        >>> def custom_converter(batch, device):
+        >>>     assert isinstance(device, chainer.backend.Device)
+        >>>     # do something with batch...
+        >>>     return device.send(batch)
+
+    This decorator puts a mark on the target converter function so that
+    Chainer can recognize that it accepts :class:`chainer.backend.Device` as
+    the device argument. For backward compatibility, the decorator also wraps
+    the function so that if the converter is called with the device argument
+    with ``int`` type, it is converted to a :class:`chainer.backend.Device`
+    instance before calling the original function. The ``int`` value indicates
+    the CUDA device of the cupy backend.
+
+    Without the decorator, the converter cannot support ChainerX devices.
+    If the batch were requested to be converted to ChainerX with such
+    converters, :class:`RuntimeError` will be raised.
+
+    """
+
+    def wrap(func):
+        func.__converter_protocol_version = 1
+
+        @functools.wraps(func)
+        def wrap_call(*args, **kwargs):
+            # Normalize the 'device' argument
+            if len(args) >= 2:
+                # specified as a positional argument
+                args = list(args)
+                args[1] = _get_device(args[1])
+            elif 'device' in kwargs:
+                kwargs['device'] = _get_device(kwargs['device'])
+            return func(*args, **kwargs)
+
+        return wrap_call
+
+    return wrap
+
+
+def _call_converter(converter, batch, device):
+    # Calls the converter.
+    # Converter can be either new-style (accepts chainer.backend.Device) or
+    # old-style (accepts int as device).
+    assert device is None or isinstance(device, backend.Device)
+
+    if getattr(converter, '__converter_protocol_version', 0) == 1:
```

Comment:
Perhaps the attribute name is too "formal"?

Reply:
I don't think versioning is useful here. Functions are not pickled, so there is no concern on compatibility nor mixed use of "versions". The attribute name would be made simpler, too.

Reply:
OK, changed the attribute name.